### DISCUSSION TITLE
Tutanota desktop App fails to build on Linux ARM64

### DISCUSSION THREAD

**Discussion by ghost** [tutao/tutanota#4439] [General]
**Describe the bug**
The Tutanota Desktop app fails to build on Linux ARM64. I have verified that it does build on an x86_64 device on the same distribution (Void Linux), the same NodeJS version and the same Kernel. It looks like that node-gyp is treating warnings as errors on ARM64, while it does not on x86_64.

The same warnings I get on ARM64, I also get on x86_64 but on x86_64 node-gyp doesn't complain and the build finishes.

I have set the architecture in buildSrc/electron-package-json-template.js from `x64` to `arm64`.

I have successfully built the Tutanota desktop client on ARM64 around half a year ago. It worked flawlessly and I cannot recall seeing the warnings I see now. 

**To Reproduce**
build the tutanota desktop App on ARM64, before step `node desktop --custom-desktop-release` finishes, node-gyp throws its error.

**Expected behavior**
The tutanota desktop App builds like normal and as expected.

**Desktop (please complete the following information):**
 - OS: Void Linux glibc ARM64 (Same error appears on Gentoo ARM64 and Arch Linux ARM though)
 - Desktop Environment/Window Manager; dwm (Suckless dynamic window manager)
 - Tutanota Desktop client Version: v3.98.14

- Gentoo ARM64 installed NodeJS version: 18.6.0
- Void Linux glibc ARM64 installed NodeJS version: 16.15.1
- Void Linux glibc x86_64 installed NodeJS version: 16.15.1
- Arch Linux ARM installed NodeJS version: 18.7.0-1

**Additional context**
The Full build log of the `node desktop --custom-desktop-release` step can be seen here: https://0x0.st/oj0i.log.

I have also tried to delete the package-lock.json and rerun the step which did not help. 

On x86_64 setting the arch to arm64 in buildSrc/electron-package-json-template.js and building succeeds.
On ARM64 setting the arch to arm64 nor x64 in buildSrc/electron-package-json-template.js and building succeeds.
I have copied the builddir witth tthe arm64 binaries from my x86_64 device over to my ARM64 device. I am unable to execute the ARM64 binary due to a missing keytar from what I read: https://0x0.st/oj0X.log

I believe that I need to tell node-gyp to treat warnings as warnings and not errors or revert the responsible commit for these warnings shown at the bottom of the build log.

**Comment by charlag** [comment]
Hi
The problem is not the warnings. We use prebuilt OpenSSL with the configuration we need when building better-sqlite3-sqlcipher and we are missing the arm version so I guess it tries to use x86 one and fails to link it (unsurprisingly).
Currently we are assuming that the target `arch` is the same as current `arch`.

https://github.com/tutao/tutanota/blob/da0fa3c5fc1d8e35992305db427e3a826b398f73/buildSrc/nativeLibraryProvider.js#L156

you would need to override that to build x86 binaries on ARM (and there are probably more places). If you want to build with ARM as a target you would need to give it correct OpenSSL build in some way.

**Comment by ghost** [comment]
> better-sqlite3-sqlcipher

I see, big thanks for the reply and heads up! How difficult would it be to integrate the right OpenSSL build into the desktop client build? Is it next to "impossible"? I want to build the Tutanota-desktop client for ARM64 using my ARM64 PC.

I see now what's the issue with better-sqlite3-sqlcipher: https://0x0.st/ojDH.log

I may be able to build better-sqlite3-cipher from source and integrate it into the desktop client. 
I have no major development skills though which should make it a lot harder.

**Comment by charlag** [comment]
better-sqlite3-sqlcipher is already built from source when you try to build desktop client. The problem is that it needs OpenSSL already built.

You would need to fork our fork of better-sqlite3-sqlcipher to add arm prebuild instead of x86 prebuilt of openssl there and then it should just work

**Comment by ghost** [comment]
> better-sqlite3-sqlcipher is already built from source when you try to build desktop client. The problem is that it needs OpenSSL already built.
> 
> You would need to fork our fork of better-sqlite3-sqlcipher to add arm prebuild instead of x86 prebuilt of openssl there and then it should just work

Hey back, I sadly don't have any clue on how to further proceed, you say that no code changes in better-sqlite3-sqlcipher are necessary, right? Just building OpenSSL for ARM64 and putting it in the right place inside the better-sqlite3-sqlcipher dir? I have found this: https://github.com/skolson/sqlcipher-openssl-build. Though there is no "linuxArm64".

**Comment by charlag** [comment]
Signal's fork has ARM version as far as I remember but I might be wrong
https://github.com/signalapp/better-sqlite3

**Reply by ghost** [reply]
Thanks, that helps a lot! I just tried to build it and it now says that there is a bad tar archive: https://0x0.st/oeMT.log

I then tried to untar this and yes:

``` </> LC_ALL=C tar -xvf sqlcipher.tar.gz
tar: This does not look like a tar archive

gzip: stdin: not in gzip format
tar: Child returned status 1
tar: Error is not recoverable: exiting now
```

Update: downloading the sqlcipher.tar.gz as stand-alone file from here: https://github.com/signalapp/better-sqlite3/blob/587721adcf31e5872e0089a1bccc9a9342f46dd5/deps/sqlcipher.tar.gz fixes it and the file can be unpacked.

Update: I put the manually downloaded sqlcipher.tar.gz into  tutanota/node_modules/better-sqlite3/deps/.
The Desktop client built and works! 
Thank you so so much! 

An update-/auto-maintain shellscript of mine will follow up.

**Reply by ghost** [reply]
```
if [ $(cat ${XDG_PROJECTS_DIR}/verfiles/tutanota-ver) != $(curl --silent "https://api.github.com/repos/tutao/tutanota/releases/latest" | jq -r .tag_name | tr -d 'tutanotadesktopandroidiosrelease-') ]

then
        TUTANOTA_VER=$(curl --silent "https://api.github.com/repos/tutao/tutanota/releases/latest" | jq -r .tag_name | tr -d 'tutanotadesktopandroidiosrelease-')

        cd ${XDG_PROJECTS_DIR}/tutanota/
        git pull; npm ci;
        
        #weirdly the client get's built fine when the commented out steps at the bottom have been done
        #after those steps the client is also buildable using just the following non-commented steps:

        cd node_modules/better-sqlite3/deps
        rm -r sqlite3.tar.gz;
        wget https://github.com/signalapp/better-sqlite3/raw/587721adcf31e5872e0089a1bccc9a9342f46dd5/deps/sqlcipher.tar.gz
        cd ../../../
       
        # There is an "unexpected error" that is shown by the client though, when only building with this method and not with the above afterwards.
        
        #cd node_modules
        #rm -r better-sqlite3;
        #git clone https://github.com/signalapp/better-sqlite3
        #cd better-sqlite3/deps; rm -r sqlcipher.tar.gz
        #wget https://github.com/signalapp/better-sqlite3/raw/587721adcf31e5872e0089a1bccc9a9342f46dd5/deps/sqlcipher.tar.gz
        #cd ../../../


        npm run build-packages;
        node desktop --custom-desktop-release;

        echo ${TUTANOTA_VER} > ${XDG_PROJECTS_DIR}/verfiles/tutanota-ver
fi

```

**Reply by charlag** [reply]
I am glad it is working for you, thank you for the update!
